File: backend/app/schemas/catalog.py
from decimal import Decimal

from pydantic import BaseModel, ConfigDict


class BrandOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    name: str
    slug: str
    logo_url: str | None = None


class CategoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    name: str
    slug: str
    path: str
    depth: int
    image_url: str | None = None
    icon: str | None = None


class CategoryTreeNode(CategoryOut):
    children: list["CategoryTreeNode"] = []


class ProductImageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    url: str
    alt_text: str | None = None
    is_primary: bool


class ProductVariantOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    sku: str
    size: str | None = None
    color: str | None = None
    color_hex: str | None = None
    price: Decimal
    mrp: Decimal
    available: int
    is_active: bool


class ProductCardOut(BaseModel):
    """Trimmed payload for listing grids — spec §17.3 ProductCard."""

    model_config = ConfigDict(from_attributes=True)

    slug: str
    title: str
    brand: str
    thumbnail_url: str | None
    price: Decimal
    mrp: Decimal
    discount_percent: Decimal
    rating_avg: Decimal | None
    rating_count: int
    currency: str
    in_stock: bool


class ProductDetailOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    slug: str
    title: str
    subtitle: str | None
    description: str | None
    brand: BrandOut
    category: CategoryOut
    gender: str | None
    material: str | None
    base_color: str | None
    currency: str
    price: Decimal
    mrp: Decimal
    discount_percent: Decimal
    rating_avg: Decimal | None
    rating_count: int
    review_count: int
    images: list[ProductImageOut]
    variants: list[ProductVariantOut]


class ProductListMeta(BaseModel):
    page: int
    per_page: int
    total: int
    total_pages: int
    has_next: bool
    search_fallback: bool = False


class FacetBucket(BaseModel):
    value: str
    count: int


class CategoryFacetBucket(BaseModel):
    slug: str
    name: str
    count: int
    is_active: bool = False


class ProductFacets(BaseModel):
    brands: list[FacetBucket] = []
    categories: list[CategoryFacetBucket] = []
    sizes: list[FacetBucket] = []
    colors: list[FacetBucket] = []
    price_range: dict[str, Decimal] | None = None


class ProductListOut(BaseModel):
    data: list[ProductCardOut]
    meta: ProductListMeta
    facets: ProductFacets

File: frontend/src/types/catalog.ts
export interface ProductCard {
  slug: string
  title: string
  brand: string
  thumbnail_url: string | null
  price: string
  mrp: string
  discount_percent: string
  rating_avg: string | null
  rating_count: number
  currency: string
  in_stock: boolean
}

export interface FacetBucket {
  value: string
  count: number
}

export interface CategoryFacetBucket {
  slug: string
  name: string
  count: number
  is_active: boolean
}

export interface ProductFacets {
  brands: FacetBucket[]
  categories: CategoryFacetBucket[]
  sizes: FacetBucket[]
  colors: FacetBucket[]
  price_range: { min: string; max: string } | null
}

export interface ProductListMeta {
  page: number
  per_page: number
  total: number
  total_pages: number
  has_next: boolean
  search_fallback: boolean
  request_id?: string
  timestamp?: string
}

export interface ProductListResponse {
  data: ProductCard[]
  meta: ProductListMeta
  facets: ProductFacets
}

export interface ProductVariant {
  id: number
  sku: string
  size: string | null
  color: string | null
  color_hex: string | null
  price: string
  mrp: string
  available: number
  is_active: boolean
}

export interface ProductImage {
  url: string
  url_webp: string | null
  blurhash: string | null
  alt_text: string | null
  is_primary: boolean
}

export interface ProductDetail {
  slug: string
  title: string
  subtitle: string | null
  description: string | null
  brand: { name: string; slug: string; logo_url: string | null }
  category: { name: string; slug: string; path: string; depth: number }
  gender: string | null
  material: string | null
  base_color: string | null
  currency: string
  price: string
  mrp: string
  discount_percent: string
  rating_avg: string | null
  rating_count: number
  review_count: number
  images: ProductImage[]
  variants: ProductVariant[]
}

export interface Category {
  name: string
  slug: string
  path: string
  depth: number
  image_url: string | null
  icon: string | null
}

export interface Brand {
  name: string
  slug: string
  logo_url: string | null
}

export interface Envelope<T> {
  data: T
  meta: { request_id: string; timestamp: string }
}
